### PR TITLE
Feature/optimize h2 use stream

### PR DESCRIPTION
**Pull request by XIEZHENGYAO**
### Issues associated with this PR

https://github.com/mosn/mosn/issues/1751

### Solutions
serverStreamConnection在handleFrame的函数中，判断第一个data帧是否是endStream，如果是，即使配置了流式也不用，如果不是，再走流式。这样可以避免大部分unary请求的损失，因为大部分unary请求data不超过一帧（16kb）.

### UT result

>  ✘ xzy@xzydeMacBook-Pro  go test -v -gcflags "all=-N -l"                                                 
2021-08-25 18:25:05,264 [INFO] register a new handler maker, name is default, is default: true
2021-08-25 18:25:05,265 [INFO] [config] processor added to configParsedCBMaps
2021-08-25 18:25:05,265 [INFO] [network] [ register pool factory] register protocol: Http2 factory
=== RUN   TestDirectResponse
--- PASS: TestDirectResponse (0.00s)
=== RUN   TestStreamConfigHandler
=== RUN   TestStreamConfigHandler/test_stream_config
=== RUN   TestStreamConfigHandler/test_invalid_default
--- PASS: TestStreamConfigHandler (0.00s)
    --- PASS: TestStreamConfigHandler/test_stream_config (0.00s)
    --- PASS: TestStreamConfigHandler/test_invalid_default (0.00s)
=== RUN   TestServerH2UseStream
--- PASS: TestServerH2UseStream (0.00s)
=== RUN   TestClientH2UseStream
--- PASS: TestClientH2UseStream (0.00s)
=== RUN   Test_get_prefixProtocolVar
--- PASS: Test_get_prefixProtocolVar (0.00s)
=== RUN   Test_get_scheme
--- PASS: Test_get_scheme (0.00s)
PASS
ok      mosn.io/mosn/pkg/stream/http2   0.444s

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result
 xzy@xzydeMacBook-Pro  ~/Desktop/projects/gopath/src/github.com/XIEZHENGYAO/mosn   feature/optimize_h2_use_stream ●✚  golint pkg/stream/http2/stream.go 
pkg/stream/http2/stream.go:84:10: if block ends with a return statement, so drop this else and outdent its block
pkg/stream/http2/stream.go:151:6: exported type StreamConfig should have comment or be unexported
pkg/stream/http2/stream.go:152:2: struct field Http2UseStream should be HTTP2UseStream

### CODE REVIEW

File: pkg/stream/http2/stream.go [hunk 1]
```diff
@@ -721,6 +730,9 @@ func (conn *clientStreamConnection) NewStream(ctx context.Context, receiver type
 	stream.sc = conn
 	stream.receiver = receiver
 	stream.conn = conn.conn
+	if useStream, ok := mosnctx.Get(ctx, types.ContextKeyUseStream).(bool); ok {
```
Comment (taoyuanyuan) [inline pkg/stream/http2/stream.go:733]: types.ContextKeyUseStream这个是没有必要的，client也可以通过同样的规则知道是否开启stream

File: pkg/stream/http2/stream.go [hunk 2]
```diff
@@ -445,6 +457,7 @@ func (conn *serverStreamConnection) onNewStreamDetect(ctx context.Context, h2s *
 	stream.ctx = mosnctx.WithValue(ctx, types.ContextKeyStreamID, stream.id)
 	stream.sc = conn
 	stream.h2s = h2s
+	// for server stream response
```
Comment (taoyuanyuan) [inline pkg/stream/http2/stream.go:460]: repsonse 的流式也使用这个优化比较好，不然会让人疑惑，只优化一半

File: pkg/stream/http2/stream.go [hunk 3]
```diff
@@ -1014,6 +1026,11 @@ func (s *clientStream) endStream() {
 	// send body and trailer
 	_, err = s.sc.protocol.Encode(s.ctx, s.h2s)
 	if err == nil {
+		if s.useStream {
```
Comment (taoyuanyuan) [inline pkg/stream/http2/stream.go:1029]: 在发送Body之前就应该判断关闭重试哟

File: pkg/stream/http2/stream.go [hunk 4]
```diff
@@ -390,6 +401,11 @@ func (conn *serverStreamConnection) handleFrame(ctx context.Context, i interface
 		}
 	}
 
+	if stream == nil {
+		log.Proxy.Errorf(ctx, "http2 server invaild steamID :%v", f)
```
Comment (taoyuanyuan) [inline pkg/stream/http2/stream.go:405]: f -> id

**Review by taoyuanyuan — APPROVED** [review]
lgtm

File: pkg/proxy/retrystate.go [hunk 5]
```diff
@@ -27,6 +27,7 @@ import (
 )
 
 type retryState struct {
+	disable          bool
```
Comment (nejisama) [inline pkg/proxy/retrystate.go:30]: 这里加这个是不是没有必要，本身retry的判断参数ctx 就是可以传递信息的，在代码实现里也是靠ctx里获取的retry state 再来修改，是不是直接在ctx里设置一个 不允许重试一类的就可以了

Comment (nejisama) [reply pkg/proxy/retrystate.go:30]: 可以用变量

File: etc/script/report.sh [hunk 6]
```diff
@@ -5,7 +5,7 @@ echo "" > coverage.txt
 
 for d in $(go list ./pkg/...  | grep -v pkg/networkextention); do
     echo "--------Run test package: $d"
-    GO111MODULE=off go test -gcflags=-l -v -coverprofile=profile.out -covermode=atomic $d
+    GO111MODULE=off go test -gcflags="all=-N -l" -v -coverprofile=profile.out -covermode=atomic $d
```
Comment (nejisama) [inline etc/script/report.sh:8]: 这里是希望禁止什么优化？

**Review by nejisama — COMMENTED** [review]
另外可以考虑补一个集成测试，就按照类似issue里描述的场景就可以

File: pkg/proxy/retrystate.go (hunk 5 above)
Comment (XIEZHENGYAO) [reply pkg/proxy/retrystate.go:30]: OK，我改一下

**Comment by XIEZHENGYAO** [comment]
> 另外可以考虑补一个集成测试，就按照类似issue里描述的场景就可以

Done.

File: pkg/stream/http2/var.go [hunk 7]
```diff
@@ -35,6 +35,9 @@ var (
 
 	builtinVariables = []variable.Variable{
 		variable.NewStringVariable(types.VarHttp2RequestScheme, nil, schemeGetter, nil, 0),
+		variable.NewVariable(types.VarHttp2RequestUseStream, nil, nil, variable.DefaultSetter, 0),
+		variable.NewVariable(types.VarHttp2ResponseUseStream, nil, nil, variable.DefaultSetter, 0),
+		variable.NewVariable(types.VarProxyDisableRetry, nil, nil, variable.DefaultSetter, 0),
```
Comment (taoyuanyuan) [inline pkg/stream/http2/var.go:40]: 放到 proxy的var里面去

Comment (XIEZHENGYAO) [reply pkg/stream/http2/var.go]: Done.

**Review by taoyuanyuan — DISMISSED** [review]
lgtm

**Review by nejisama — APPROVED** [review]

**Review by taoyuanyuan — APPROVED** [review]
lgtm

**Pull request merged**